src/core/agents.jl

```diff
@@ -57,60 +71,168 @@ mutable struct Person{T} <: AbstractAgent
+type-annotate `x::Animal` in `f`. Don't annotate any type. Annotating a type
+only becomes useful if there are at least two "abstract" groups, like `Animal, Person`.
+Then it would make sense to define
+```julia
+Person = Union{Fisher, Baker}
+f(x::Animal) = ... # uses `CommonTraits` fields
+f(x::Person) = ... # uses fields that all "persons" have
+```
 """
-    ContinuousAgent{D}
-Combine with [`@agent`](@ref) to create an agent type for `D`-dimensional
-[`ContinuousSpace`](@ref). It attributes the fields
-`id::Int, pos::NTuple{D,Float64}, vel::NTuple{D,Float64}`
-to the start of the agent type.
-"""
-mutable struct ContinuousAgent{D} <: AbstractAgent
-    id::Int
-    pos::NTuple{D,Float64}
-    vel::NTuple{D,Float64}
+macro agent(new_name, base_type, extra_fields)
+    # This macro was generated with the guidance of @rdeits on Discourse:
+    # https://discourse.julialang.org/t/
+    # metaprogramming-obtain-actual-type-from-symbol-for-field-inheritance/84912
+
+    # We start with a quote. All macros return a quote to be evaluated
+    quote
+        let
+            # Here we collect the field names and types from the base type
+            # Because the base type already exists, we escape the symbols to obtain it
+            base_fieldnames = fieldnames($(esc(base_type)))
+            base_fieldtypes = [t for t in getproperty($(esc(base_type)), :types)]
+            base_fields = [:($f::$T) for (f, T) in zip(base_fieldnames, base_fieldtypes)]
+            # Then, we prime the additional name and fields into QuoteNodes
+            # We have to do this to be able to interpolate them into an inner quote.
+            name = $(QuoteNode(new_name))
+            additional_fields = $(QuoteNode(extra_fields.args))
+            # Now we start an inner quote. This is because our macro needs to call `eval`
+            # However, this should never happen inside the main body of a macro
+            # There are several reasons for that, see the cited discussion at the top
+            expr = quote
+                mutable struct $name <: AbstractAgent
+                    $(base_fields...)
+                    $(additional_fields...)
+                end
+            end
+            # @show expr # uncomment this to see that the final expression looks as desired
+            # It is important to evaluate the macro in the module that it was called at
+            Core.eval($(__module__), expr)
+        end
+    end
 end
+# TODO: I do not know how to merge these two macros to remove code duplication.
+# There should be away that only the 4-argument version is used
+# and the 3-argument version just passes `AbstractAgent` to the 4-argument.
+macro agent(new_name, base_type, super_type, extra_fields)
+    # This macro was generated with the guidance of @rdeits on Discourse:
+    # https://discourse.julialang.org/t/
+    # metaprogramming-obtain-actual-type-from-symbol-for-field-inheritance/84912
```

Comment:
```suggestion
    # https://discourse.julialang.org/t/metaprogramming-obtain-actual-type-from-symbol-for-field-inheritance/84912
```

Same as above, to make this copy-pasteable